core/job/service/job_service.go

```diff
@@ -262,6 +262,89 @@ func (j *JobService) Update(ctx context.Context, jobTenant tenant.Tenant, specs
 	return job.Jobs(updatedJobs).GetJobNames(), me.ToErr()
 }
 
+func (j *JobService) Upsert(ctx context.Context, jobTenant tenant.Tenant, specs []*job.Spec) ([]job.Name, error) {
+	logWriter := writer.NewLogWriter(j.logger)
+	me := errors.NewMultiError("upsert specs errors")
+
+	tenantWithDetails, err := j.tenantDetailsGetter.GetDetails(ctx, jobTenant)
+	if err != nil {
+		j.logger.Error("error getting tenant details: %s", err)
+		return []job.Name{}, err
+	}
+
+	jobs, err := j.generateJobs(ctx, tenantWithDetails, specs, logWriter)
+	me.Append(err)
+
+	existingJobs := make(map[job.Name]*job.Job)
+	var jobsToAdd, jobsToUpdate []*job.Job
+	for _, jobToUpsert := range jobs {
+		existingJob, err := j.jobRepo.GetByJobName(ctx, jobTenant.ProjectName(), jobToUpsert.Spec().Name())
+		if err != nil {
+			if !errors.IsErrorType(err, errors.ErrNotFound) {
+				me.Append(err)
+				continue
+			}
+			jobsToAdd = append(jobsToAdd, jobToUpsert)
+			continue
+		}
+		jobsToUpdate = append(jobsToUpdate, jobToUpsert)
+		existingJobs[jobToUpsert.Spec().Name()] = existingJob
+	}
+
+	var upsertedJobs, addedJobs, updatedJobs []*job.Job
+	if len(jobsToAdd) > 0 {
+		addedJobs, err = j.jobRepo.Add(ctx, jobsToAdd)
+		upsertedJobs = append(upsertedJobs, addedJobs...)
+		logWriter.Write(writer.LogLevelInfo, fmt.Sprintf("[%s] added %d jobs", tenantWithDetails.Namespace().Name().String(), len(addedJobs)))
```

Comment:
can we omit the log if the added/updated jobs is 0?

Reply:
sure

Reply:
updated